File: py/src/apptheory/context.py
from __future__ import annotations

import json as jsonlib
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Protocol

from apptheory.aws_events import AppSyncResolverEvent
from apptheory.clock import Clock, RealClock
from apptheory.errors import AppError
from apptheory.ids import IdGenerator, RealIdGenerator
from apptheory.request import Request
from apptheory.source_provenance import SourceProvenance, normalize_source_provenance, unknown_source_provenance


@dataclass(slots=True)
class AppSyncContext:
    field_name: str
    parent_type_name: str
    arguments: dict[str, Any]
    identity: dict[str, Any]
    source: dict[str, Any]
    variables: dict[str, Any]
    stash: dict[str, Any]
    prev: Any | None
    request_headers: dict[str, str]
    raw_event: AppSyncResolverEvent | dict[str, Any]

    def __init__(
        self,
        *,
        field_name: str = "",
        parent_type_name: str = "",
        arguments: dict[str, Any] | None = None,
        identity: dict[str, Any] | None = None,
        source: dict[str, Any] | None = None,
        variables: dict[str, Any] | None = None,
        stash: dict[str, Any] | None = None,
        prev: Any | None = None,
        request_headers: dict[str, str] | None = None,
        raw_event: AppSyncResolverEvent | dict[str, Any] | None = None,
    ) -> None:
        self.field_name = str(field_name or "").strip()
        self.parent_type_name = str(parent_type_name or "").strip()
        self.arguments = dict(arguments or {})
        self.identity = dict(identity or {})
        self.source = dict(source or {})
        self.variables = dict(variables or {})
        self.stash = dict(stash or {})
        self.prev = prev
        self.request_headers = {
            str(key): str(value) for key, value in dict(request_headers or {}).items() if str(key).strip()
        }
        self.raw_event = dict(raw_event or {})


@dataclass(slots=True)
class Context:
    request: Request
    params: dict[str, str]
    clock: Clock
    id_generator: IdGenerator
    ctx: Any | None
    request_id: str
    tenant_id: str
    auth_identity: str
    remaining_ms: int
    middleware_trace: list[str]
    websocket: WebSocketContext | None
    appsync: AppSyncContext | None
    _values: dict[str, Any]

    def __init__(
        self,
        *,
        request: Request,
        params: dict[str, str] | None = None,
        clock: Clock | None = None,
        id_generator: IdGenerator | None = None,
        ctx: Any | None = None,
        request_id: str = "",
        tenant_id: str = "",
        auth_identity: str = "",
        remaining_ms: int = 0,
        middleware_trace: list[str] | None = None,
        websocket: WebSocketContext | None = None,
        appsync: AppSyncContext | None = None,
    ) -> None:
        self.request = request
        self.params = params or {}
        self.clock = clock or RealClock()
        self.id_generator = id_generator or RealIdGenerator()
        self.ctx = ctx
        self.request_id = str(request_id)
        self.tenant_id = str(tenant_id)
        self.auth_identity = str(auth_identity)
        self.remaining_ms = int(remaining_ms or 0)
        self.middleware_trace = middleware_trace if middleware_trace is not None else []
        self.websocket = websocket
        self.appsync = appsync
        self._values = {}

    def now(self):
        return self.clock.now()

    def new_id(self) -> str:
        return self.id_generator.new_id()

    def param(self, name: str) -> str:
        return self.params.get(name, "")

    def set(self, key: str, value: Any) -> None:
        name = str(key or "").strip()
        if not name:
            return
        self._values[name] = value

    def get(self, key: str, default: Any | None = None) -> Any:
        name = str(key or "").strip()
        if not name:
            return default
        return self._values.get(name, default)

    def source_provenance(self) -> SourceProvenance:
        if self is None:
            return unknown_source_provenance()
        return normalize_source_provenance(self.request.source_provenance)

    def source_ip(self) -> str:
        return self.source_provenance().source_ip

    def as_websocket(self) -> WebSocketContext | None:
        return self.websocket

    def as_appsync(self) -> AppSyncContext | None:
        return self.appsync

    def json_value(self) -> Any:
        if not _has_json_content_type(self.request.headers):
            raise AppError("app.bad_request", "invalid json")
        if not self.request.body:
            return None
        try:
            return jsonlib.loads(self.request.body.decode("utf-8"))
        except Exception:  # noqa: BLE001
            raise AppError("app.bad_request", "invalid json") from None


@dataclass(slots=True)
class EventContext:
    clock: Clock
    id_generator: IdGenerator
    ctx: Any | None
    request_id: str
    remaining_ms: int
    _values: dict[str, Any]

    def __init__(
        self,
        *,
        clock: Clock | None = None,
        id_generator: IdGenerator | None = None,
        ctx: Any | None = None,
        request_id: str = "",
        remaining_ms: int = 0,
    ) -> None:
        self.clock = clock or RealClock()
        self.id_generator = id_generator or RealIdGenerator()
        self.ctx = ctx
        self.request_id = str(request_id)
        self.remaining_ms = int(remaining_ms or 0)
        self._values = {}

    def now(self):
        return self.clock.now()

    def new_id(self) -> str:
        return self.id_generator.new_id()

    def set(self, key: str, value: Any) -> None:
        name = str(key or "").strip()
        if not name:
            return
        self._values[name] = value

    def get(self, key: str, default: Any | None = None) -> Any:
        name = str(key or "").strip()
        if not name:
            return default
        return self._values.get(name, default)


class WebSocketManagementClient(Protocol):
    def post_to_connection(self, connection_id: str, data: bytes) -> Any: ...

    def get_connection(self, connection_id: str) -> Any: ...

    def delete_connection(self, connection_id: str) -> Any: ...


WebSocketClientFactory = Callable[[str, Any | None], WebSocketManagementClient]


@dataclass(slots=True)
class WebSocketContext:
    clock: Clock
    id_generator: IdGenerator
    ctx: Any | None
    request_id: str
    remaining_ms: int
    connection_id: str
    route_key: str
    domain_name: str
    stage: str
    event_type: str
    management_endpoint: str
    body: bytes
    client_factory: WebSocketClientFactory | None
    _client: WebSocketManagementClient | None
    _client_error: Exception | None

    def __init__(
        self,
        *,
        clock: Clock | None = None,
        id_generator: IdGenerator | None = None,
        ctx: Any | None = None,
        request_id: str = "",
        remaining_ms: int = 0,
        connection_id: str = "",
        route_key: str = "",
        domain_name: str = "",
        stage: str = "",
        event_type: str = "",
        management_endpoint: str = "",
        body: bytes | None = None,
        client_factory: WebSocketClientFactory | None = None,
    ) -> None:
        self.clock = clock or RealClock()
        self.id_generator = id_generator or RealIdGenerator()
        self.ctx = ctx
        self.request_id = str(request_id)
        self.remaining_ms = int(remaining_ms or 0)
        self.connection_id = str(connection_id or "").strip()
        self.route_key = str(route_key or "").strip()
        self.domain_name = str(domain_name or "").strip()
        self.stage = str(stage or "").strip()
        self.event_type = str(event_type or "").strip()
        self.management_endpoint = str(management_endpoint or "").strip()
        self.body = bytes(body or b"")
        self.client_factory = client_factory
        self._client = None
        self._client_error = None

    def now(self):
        return self.clock.now()

    def new_id(self) -> str:
        return self.id_generator.new_id()

    def _management_client(self) -> WebSocketManagementClient:
        if self._client is not None:
            return self._client
        if self._client_error is not None:
            raise self._client_error
        if self.client_factory is None:
            self._client_error = RuntimeError("apptheory: missing websocket client factory")
            raise self._client_error

        try:
            client = self.client_factory(self.management_endpoint, self.ctx)
        except Exception as exc:
            self._client_error = exc
            raise

        if client is None:
            self._client_error = RuntimeError("apptheory: websocket client factory returned None")
            raise self._client_error

        self._client = client
        return client

    def send_message(self, data: bytes) -> None:
        if not self.connection_id:
            raise RuntimeError("apptheory: websocket connection id is empty")
        client = self._management_client()
        client.post_to_connection(self.connection_id, bytes(data or b""))

    def send_json_message(self, value: Any) -> None:
        payload = jsonlib.dumps(value, separators=(",", ":"), ensure_ascii=False, sort_keys=True).encode("utf-8")
        self.send_message(payload)


def _has_json_content_type(headers: dict[str, list[str]]) -> bool:
    for value in headers.get("content-type", []):
        v = str(value).strip().lower()
        if v.startswith("application/json"):
            return True
    return False

File: ts/src/context.ts
import { Buffer } from "node:buffer";

import type { AppSyncResolverEvent } from "./aws-types.js";
import { RealClock, type Clock } from "./clock.js";
import { AppError } from "./errors.js";
import { RandomIdGenerator, type IdGenerator } from "./ids.js";
import { toBuffer } from "./internal/http.js";
import { hasJSONContentType } from "./internal/response.js";
import { normalizeSourceProvenance } from "./internal/source-provenance.js";
import type { Headers, Query, Response, SourceProvenance } from "./types.js";

export interface WebSocketManagementClientLike {
  postToConnection: (
    connectionId: string,
    data: Uint8Array,
  ) => void | Promise<void>;
  getConnection: (connectionId: string) => unknown | Promise<unknown>;
  deleteConnection: (connectionId: string) => void | Promise<void>;
}

export type WebSocketClientFactory = (
  endpoint: string,
  ctx: unknown | null,
) => WebSocketManagementClientLike | Promise<WebSocketManagementClientLike>;

export class AppSyncContext {
  readonly fieldName: string;
  readonly parentTypeName: string;
  readonly arguments: Record<string, unknown>;
  readonly identity: Record<string, unknown>;
  readonly source: Record<string, unknown>;
  readonly variables: Record<string, unknown>;
  readonly stash: Record<string, unknown>;
  readonly prev: unknown;
  readonly requestHeaders: Record<string, string>;
  readonly rawEvent: AppSyncResolverEvent;

  constructor(options: {
    fieldName?: string;
    parentTypeName?: string;
    arguments?: Record<string, unknown>;
    identity?: Record<string, unknown>;
    source?: Record<string, unknown>;
    variables?: Record<string, unknown>;
    stash?: Record<string, unknown>;
    prev?: unknown;
    requestHeaders?: Record<string, string>;
    rawEvent: AppSyncResolverEvent;
  }) {
    this.fieldName = String(options.fieldName ?? "").trim();
    this.parentTypeName = String(options.parentTypeName ?? "").trim();
    this.arguments =
      options.arguments && typeof options.arguments === "object"
        ? { ...options.arguments }
        : {};
    this.identity =
      options.identity && typeof options.identity === "object"
        ? { ...options.identity }
        : {};
    this.source =
      options.source && typeof options.source === "object"
        ? { ...options.source }
        : {};
    this.variables =
      options.variables && typeof options.variables === "object"
        ? { ...options.variables }
        : {};
    this.stash =
      options.stash && typeof options.stash === "object"
        ? { ...options.stash }
        : {};
    this.prev = options.prev ?? null;
    this.requestHeaders =
      options.requestHeaders && typeof options.requestHeaders === "object"
        ? { ...options.requestHeaders }
        : {};
    const rawInfo = options.rawEvent.info ?? {};
    this.rawEvent = {
      ...options.rawEvent,
      arguments:
        options.rawEvent.arguments &&
        typeof options.rawEvent.arguments === "object"
          ? { ...options.rawEvent.arguments }
          : {},
      identity:
        options.rawEvent.identity &&
        typeof options.rawEvent.identity === "object"
          ? { ...options.rawEvent.identity }
          : {},
      source:
        options.rawEvent.source && typeof options.rawEvent.source === "object"
          ? { ...options.rawEvent.source }
          : {},
      stash:
        options.rawEvent.stash && typeof options.rawEvent.stash === "object"
          ? { ...options.rawEvent.stash }
          : {},
      ...(options.rawEvent.request &&
      typeof options.rawEvent.request === "object" &&
      options.rawEvent.request.headers &&
      typeof options.rawEvent.request.headers === "object"
        ? { request: { headers: { ...options.rawEvent.request.headers } } }
        : {}),
      info: {
        ...rawInfo,
        ...(rawInfo.variables && typeof rawInfo.variables === "object"
          ? { variables: { ...rawInfo.variables } }
          : {}),
        ...(Array.isArray(rawInfo.selectionSetList)
          ? { selectionSetList: [...rawInfo.selectionSetList] }
          : {}),
      },
    };
  }
}

export class Context {
  readonly ctx: unknown | null;
  readonly request: {
    method: string;
    path: string;
    query: Query;
    headers: Headers;
    cookies: Record<string, string>;
    body: Uint8Array;
    isBase64: boolean;
    sourceProvenance: SourceProvenance;
  };
  readonly params: Record<string, string>;

  requestId: string;
  tenantId: string;
  authIdentity: string;
  remainingMs: number;
  middlewareTrace: string[];

  private readonly _clock: Clock;
  private readonly _ids: IdGenerator;
  private readonly _webSocket: WebSocketContext | null;
  private readonly _appSync: AppSyncContext | null;
  private readonly _values: Map<string, unknown>;

  constructor(options: {
    request: Context["request"];
    params?: Record<string, string>;
    clock?: Clock;
    ids?: IdGenerator;
    ctx?: unknown;
    requestId?: string;
    tenantId?: string;
    authIdentity?: string;
    remainingMs?: number;
    middlewareTrace?: string[];
    webSocket?: WebSocketContext | null;
    appSync?: AppSyncContext | null;
  }) {
    this.ctx = options.ctx ?? null;
    this.request = options.request;
    this.params = options.params ?? {};
    this._clock = options.clock ?? new RealClock();
    this._ids = options.ids ?? new RandomIdGenerator();
    this.requestId = options.requestId ?? "";
    this.tenantId = options.tenantId ?? "";
    this.authIdentity = options.authIdentity ?? "";
    this.remainingMs = Number(options.remainingMs ?? 0);
    this.middlewareTrace = Array.isArray(options.middlewareTrace)
      ? options.middlewareTrace
      : [];
    this._webSocket = options.webSocket ?? null;
    this._appSync = options.appSync ?? null;
    this._values = new Map();
  }

  now(): Date {
    return this._clock.now();
  }

  newId(): string {
    return this._ids.newId();
  }

  param(name: string): string {
    return this.params[String(name)] ?? "";
  }

  set(key: string, value: unknown): void {
    const k = String(key ?? "").trim();
    if (!k) return;
    this._values.set(k, value);
  }

  get(key: string): unknown {
    const k = String(key ?? "").trim();
    if (!k) return undefined;
    return this._values.get(k);
  }

  sourceProvenance(): SourceProvenance {
    return normalizeSourceProvenance(this.request.sourceProvenance);
  }

  sourceIP(): string {
    return this.sourceProvenance().sourceIP;
  }

  jsonValue(): unknown {
    if (!hasJSONContentType(this.request.headers)) {
      throw new AppError("app.bad_request", "invalid json");
    }
    if (this.request.body.length === 0) {
      return null;
    }
    try {
      return JSON.parse(Buffer.from(this.request.body).toString("utf8"));
    } catch {
      throw new AppError("app.bad_request", "invalid json");
    }
  }

  asWebSocket(): WebSocketContext | null {
    return this._webSocket;
  }

  asAppSync(): AppSyncContext | null {
    return this._appSync;
  }
}

export type Handler = (ctx: Context) => Response | Promise<Response>;

export type Middleware = (
  ctx: Context,
  next: Handler,
) => Response | Promise<Response>;

export class EventContext {
  readonly ctx: unknown | null;
  requestId: string;
  remainingMs: number;

  private readonly _clock: Clock;
  private readonly _ids: IdGenerator;
  private readonly _values: Map<string, unknown>;

  constructor(options: {
    clock?: Clock;
    ids?: IdGenerator;
    ctx?: unknown;
    requestId?: string;
    remainingMs?: number;
  }) {
    this.ctx = options.ctx ?? null;
    this._clock = options.clock ?? new RealClock();
    this._ids = options.ids ?? new RandomIdGenerator();
    this.requestId = options.requestId ?? "";
    this.remainingMs = Number(options.remainingMs ?? 0);
    this._values = new Map();
  }

  now(): Date {
    return this._clock.now();
  }

  newId(): string {
    return this._ids.newId();
  }

  set(key: string, value: unknown): void {
    const k = String(key ?? "").trim();
    if (!k) return;
    this._values.set(k, value);
  }

  get(key: string): unknown {
    const k = String(key ?? "").trim();
    if (!k) return undefined;
    return this._values.get(k);
  }
}

export type EventHandler = (
  ctx: EventContext,
  event: unknown,
) => unknown | Promise<unknown>;

export type EventMiddleware = (
  ctx: EventContext,
  event: unknown,
  next: () => unknown | Promise<unknown>,
) => unknown | Promise<unknown>;

export class WebSocketContext {
  readonly ctx: unknown | null;
  requestId: string;
  remainingMs: number;

  connectionId: string;
  routeKey: string;
  domainName: string;
  stage: string;
  eventType: string;
  managementEndpoint: string;
  body: Uint8Array;

  private readonly _clock: Clock;
  private readonly _ids: IdGenerator;
  private readonly _clientFactory: WebSocketClientFactory | null;
  private _client: WebSocketManagementClientLike | null;
  private _clientError: Error | null;

  constructor(options: {
    clock?: Clock;
    ids?: IdGenerator;
    ctx?: unknown;
    requestId?: string;
    remainingMs?: number;
    connectionId?: string;
    routeKey?: string;
    domainName?: string;
    stage?: string;
    eventType?: string;
    managementEndpoint?: string;
    body?: Uint8Array | Buffer | string | null;
    clientFactory?: WebSocketClientFactory | null;
  }) {
    this.ctx = options.ctx ?? null;
    this._clock = options.clock ?? new RealClock();
    this._ids = options.ids ?? new RandomIdGenerator();
    this.requestId = options.requestId ?? "";
    this.remainingMs = Number(options.remainingMs ?? 0);

    this.connectionId = String(options.connectionId ?? "").trim();
    this.routeKey = String(options.routeKey ?? "").trim();
    this.domainName = String(options.domainName ?? "").trim();
    this.stage = String(options.stage ?? "").trim();
    this.eventType = String(options.eventType ?? "").trim();
    this.managementEndpoint = String(options.managementEndpoint ?? "").trim();
    this.body = toBuffer(options.body);

    this._clientFactory =
      typeof options.clientFactory === "function"
        ? options.clientFactory
        : null;
    this._client = null;
    this._clientError = null;
  }

  now(): Date {
    return this._clock.now();
  }

  newId(): string {
    return this._ids.newId();
  }

  private async _managementClient(): Promise<WebSocketManagementClientLike> {
    if (this._client || this._clientError) {
      if (this._clientError) throw this._clientError;
      return this._client as WebSocketManagementClientLike;
    }

    if (!this._clientFactory) {
      this._clientError = new Error(
        "apptheory: missing websocket client factory",
      );
      throw this._clientError;
    }

    const client = await this._clientFactory(this.managementEndpoint, this.ctx);
    if (!client) {
      this._clientError = new Error(
        "apptheory: websocket client factory returned null",
      );
      throw this._clientError;
    }

    this._client = client;
    return client;
  }

  async sendMessage(data: Uint8Array): Promise<void> {
    const id = String(this.connectionId ?? "").trim();
    if (!id) throw new Error("apptheory: websocket connection id is empty");

    const client = await this._managementClient();
    if (typeof client.postToConnection !== "function") {
      throw new Error("apptheory: websocket client missing postToConnection");
    }

    await client.postToConnection(id, toBuffer(data));
  }

  async sendJSONMessage(value: unknown): Promise<void> {
    await this.sendMessage(Buffer.from(JSON.stringify(value), "utf8"));
  }
}
